### PR TITLE
Custom pipeline 

### PR DESCRIPTION
**Pull request by sklam**
This exposes a `pipeline_class` kwarg in `@jit` and `@generated_jit` for user to provide a custom compiler pipeline.  Valid value to `pipeline_class` is a subclass of `numba.compiler.BasePipeline`.  It is expected that the `.define_pipelines()` method is override.  For example usage, see definition of `numba.compiler.Pipeline`, the default compiler pipeline.

TODOS:

- [x] docs

### CODE REVIEW

**Comment by ehsantn** [comment]
@sklam I just tried this patch for my use case and it works very well.

File: docs/source/developer/custom_pipeline.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
+.. _arch-pipeline:
+
+========================
+Notes on Custom Pipeline
```
Comment (stuartarchibald) [inline docs/source/developer/custom_pipeline.rst]: s/Pipeline/Pipelines/

File: docs/source/developer/custom_pipeline.rst [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+========================
+
+.. warning:: The custom pipeline feature is regard as expert only.  Modifying
```
Comment (stuartarchibald) [inline docs/source/developer/custom_pipeline.rst]: s/regard as expert only/regarded as for being for expert use only/

File: docs/source/developer/custom_pipeline.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+             the compiler behavior can invalidate internal assumptions in the
+             numba source code.
+
+
+For library developers looking for a way to extend or modify the compiler
+behavior, you can do so by defining a custom compiler pipeline by inheriting
+from ``numba.compiler.BasePipeline``.  The default numba pipeline is defined
+as ``numba.compiler.Pipeline``.  It implements the ``.define_pipelines()``
```
Comment (stuartarchibald) [inline docs/source/developer/custom_pipeline.rst]: s/. I/, i/

File: docs/source/developer/custom_pipeline.rst [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+method, which adds the *nopython-mode*, *object-mode* and *interpreted-mode*
+pipelines.  These three pipelines are defined by ``.define_nopython_pipeline``,
```
Comment (stuartarchibald) [inline docs/source/developer/custom_pipeline.rst]: Perhaps `These three pipelines are defined in ``BasePipeline`` by the methods ``.define_nopython_pipeline``, ``.define_objectmode_pipeline`` and ``.define_interpreted_pipeline``,
respectively.` ?

File: numba/compiler.py [hunk 5]
```diff
@@ -613,21 +622,22 @@ def _backend(self, lowerfn, objectmode):
...
-                                 call_helper=lowered.call_helper,
-                                 signature=signature,
-                                 objectmode=objectmode,
-                                 interpmode=False,
-                                 lifted=self.lifted,
-                                 fndesc=lowered.fndesc,
-                                 environment=lowered.env,
-                                 has_dynamic_globals=lowered.has_dynamic_globals,
-                                 )
+        self.cr = compile_result(
```
Comment (stuartarchibald) [inline numba/compiler.py]: There seems to be a lot of formatting churn?! autopep8?

File: numba/compiler.py [hunk 6]
```diff
@@ -680,48 +694,94 @@ def stage_cleanup(self):
...
 
-    def _compile_core(self):
+    def define_pipelines(self, pm):
+        """Child classes override this to customize the pipeline.
         """
-        Populate and run compiler pipeline
+        raise NotImplementedError()
+
+    def add_preprocessing_stage(self, pm):
+        """Add the preprocessing stage that analyze the bytecode to prepare
```
Comment (stuartarchibald) [inline numba/compiler.py]: s/analyze/analyzes/

File: numba/compiler.py [hunk 7]
```diff
@@ -680,48 +694,94 @@ def stage_cleanup(self):
...
-            if self.flags.auto_parallel.enabled:
-                pm.add_stage(self.stage_pre_parfor_pass, "Preprocessing for parfors")
-            if not self.flags.no_rewrites:
-                pm.add_stage(self.stage_nopython_rewrites, "nopython rewrites")
-            if self.flags.auto_parallel.enabled:
-                pm.add_stage(self.stage_parfor_pass, "convert to parfors")
-            pm.add_stage(self.stage_nopython_backend, "nopython mode backend")
-            pm.add_stage(self.stage_cleanup, "cleanup intermediate results")
+    def add_pre_typing_stage(self, pm):
+        """Add any stages that goes before type-inference.
```
Comment (stuartarchibald) [inline numba/compiler.py]: s/goes/go/

File: numba/compiler.py [hunk 8]
```diff
@@ -680,48 +694,94 @@ def stage_cleanup(self):
...
+        The current stages contains type-agnostic rewrite passes.
```
Comment (stuartarchibald) [inline numba/compiler.py]: s/contains/contain/

File: numba/compiler.py [hunk 9]
```diff
@@ -680,48 +694,94 @@ def stage_cleanup(self):
...
+        self.add_optimization_stage(pm)
+        self.add_lowering_stage(pm)
+        self.add_cleanup_stage(pm)
+
+    def define_objectmode_pipeline(self, pm, name='object'):
+        """Add the object-mode pipeline to the pipeline manager
+        """
+        pm.create_pipeline(name)
+        self.add_preprocessing_stage(pm)
+        pm.add_stage(self.stage_objectmode_frontend,
```
Comment (stuartarchibald) [inline numba/compiler.py]: The inversion of concern here is a bit weird, `self.add_stage_foo(pm)` vs. `pm.add_stage(foo)`. Perhaps another refactor at a later date?

File: numba/compiler.py [hunk 10]
```diff
@@ -746,6 +806,18 @@ def _compile_ir(self):
         return self._compile_core()
 
 
+class Pipeline(BasePipeline):
+    """The default compiler pipeline
+    """
+    def define_pipelines(self, pm):
```
Comment (stuartarchibald) [inline numba/compiler.py]: This is so much neater :)

File: numba/compiler.py [hunk 11]
```diff
@@ -768,15 +840,29 @@ def _make_subtarget(targetctx, flags):
...
+        target context
+    func : function
+        the python function to be compiled
+    args : tuple, list
+        argument types
+    return_type :
+        Use ``None`` to indicate void return
+    flags : numba.compiler.Flags
+        compiler flags
+    library : numba.codegen.CodeLibrary
```
Comment (stuartarchibald) [inline numba/compiler.py]: Perhaps add some further explanation of this?

File: numba/decorators.py [hunk 12]
```diff
@@ -31,7 +31,8 @@ def autojit(*args, **kws):
                                  "Signatures should be passed as the first "
                                  "positional argument.")
 
-def jit(signature_or_function=None, locals={}, target='cpu', cache=False, **options):
+def jit(signature_or_function=None, locals={}, target='cpu', cache=False,
+        pipeline_class=None, **options):
```
Comment (stuartarchibald) [inline numba/decorators.py]: Need this be called `_class` ? Though I suppose it does not hurt.

File: numba/compiler.py (hunk 5 above)
Comment (sklam) [reply numba/compiler.py]: All manual =)

File: numba/compiler.py (hunk 9 above)
Comment (sklam) [reply numba/compiler.py]: I agree.  That are still a lot of things that need cleaning up.

**Comment by stuartarchibald** [comment]
Thanks for the fix ups @sklam. Merging.

**Pull request merged**